proxy_agent/src/shared_state.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+use crate::key_keeper::key::Key;
+use std::sync::{Arc, Mutex};
+
+const UNKNOWN_STATUS_MESSAGE: &str = "Status unknown.";
+
+#[derive(Clone)]
+pub struct SharedState {
+    // key_keeper
+    key: Option<Key>,
+    current_secure_channel_state: String,
+    wireserver_rule_id: String,
+    imds_rule_id: String,
+    key_keeper_shutdown: bool,
+    key_keeper_status_message: String,
+    // proxy_listener
+    proxy_listner_shutdown: bool,
+    connection_count: u128,
+    proxy_listner_status_message: String,
+    // Add more state fields as needed,
+    // keep the fields related to the same module together
+    // keep the fields as private to avoid the direct access from outside via Arc<Mutex<SharedState>>.lock().unwrap()
+    // use wrapper functions to access the state fields, it does quick release the lock
+}
+
+pub fn new_shared_state() -> Arc<Mutex<SharedState>> {
```

Comment:
It's idiomatic to implement this on the structure:

```
impl SharedState {
    pub fn new() -> Arc<Mutex<SharedState>> {
        Arc::new(Mutex::new(SharedState::default()))
    }
}
```

and then use with `SharedState::new()`